faunadb/client.py

```diff
@@ -1,14 +1,40 @@
 from time import time
 # pylint: disable=redefined-builtin
 from builtins import object
+import threading
 
 from requests import Request, Session
+from requests.auth import HTTPBasicAuth
+from requests.adapters import HTTPAdapter
 
 from faunadb.errors import _get_or_raise, FaunaError, UnexpectedError
 from faunadb.query import _wrap
 from faunadb.request_result import RequestResult
 from faunadb._json import parse_json_or_none, to_json
 
+class _Counter(object):
+  def __init__(self, init_value=0):
+    self.lock = threading.Lock()
+    self.counter = init_value
+
+  def __str__(self):
+    return "Counter(%s)" % self.counter
+
+  def increment(self):
+    with self.lock:
+      self.counter += 1
+      return self.counter
+
+  def get_and_increment(self):
```

Comment:
There is still a race condition in here:

- Thread 1 calls `decrement`: decrements counter to 0 and close the session
- Thread 2 calls `get_and_increment`: increments the counter to 1, returns 0, therefore fail to create a new session client
- Thread 3 calls `get_and_increment`: increments the counter to 2, returns 1, allowing a new session client to be create with a closed session.

The knowledge of a live session must be composed by its ref count plus a flag that tells you if the session was closed already.